man/zopeedit.1
.TH zopeedit 1 "Jul 29, 2002" "Zope External Editor"
.SH NAME
.B zopeedit
\- helper application that handles the interaction between the
Zope ExternalEditor product and the client side applications

.SH SYNOPSIS
.B This program is intended to be called by the web browser in response to the application/x-zope-edit MIME type, and is not meant to be called directly by the user except for debugging or testing purposes.

.SH DESCRIPTION
Most web browsers should respect the MIME handling instructions in
the /etc/mailcap metamail capabilities file, where an entry should
have automatically been placed by the system to let zopeedit handle
the application/x-zope-edit MIME type.

However, if your web browser does not respect mailcap, or if it was
not automatically updated during installation, you can simply specify

.B /usr/bin/zopeedit

as a new helper application in your web browser of choice to handle
downloads of MIME type

.B application/x-zope-edit

for you to be able to edit Zope objects using your favourite editor.

Further customization may be made after the first run of the helper
application on the file

.B ~/.zope-external-edit

More information on configuration and debugging are available in the
upstream author's home page at
<http://www.zope.org/Members/Caseman/ExternalEditor>.

.SH FILES
.TP
.I ~/.zope-external-edit

.SH "SEE ALSO"
.BR mailcap(5)

.SH AUTHOR
This manual page was initially written by Andreas Tille <tille@debian.org>,
and is maintained by Federico Sevilla III <jijo@free.net.ph> as part of the
zopeedit package for the Debian GNU/Linux system. It may be used freely by
others.


CHANGES.rst
Changelog
=========

4.1 (2026-03-19)
----------------

- Move all package metadata to ``pyproject.toml``.

- Add support for Python 3.12, 3.13 and 3.14.

- Drop support for Python 3.7, 3.8 and 3.9.


4.0.1 (2024-09-16)
------------------

- Correction of the use of Pdata when editing content
  (`#23 <https://github.com/zopefoundation/Products.ExternalEditor/issues/23>`_)


4.0 (2023-07-04)
----------------

- Ensure an object's editable body is cast to bytes
  (`#20 <https://github.com/zopefoundation/Products.ExternalEditor/issues/20>`_)

- Add support for Python 3.11.

- Drop support for Python 2.7, 3.5, 3.6.


3.1.0 (2022-06-24)
------------------

- Add support for Python 3.9, 3.10.

- Change package structure to move package code into a ``src`` subfolder.

- Fix bug which prevented ZMI from rendering, when Products.ExternalEdit was
  installed.
  (`#18 <https://github.com/zopefoundation/Products.ExternalEditor/pull/18>`_)


3.0.1 (2020-10-30)
------------------

Bug fixes:

- Add support for recent Python 3 versions [ale-rt]


3.0 (2019-02-16)
----------------

- Fix zmi-patches for Zope 4.
  [pbauer]

- Drop support for Python 2.6.

- Adapt tests to Zope 4.

- Add needed dependencies in `setup.py`.

- Prepare for Python 2 / 3 compatibility  [davilima6]


2.0.3 (2017-07-17)
------------------

- Document Zope version compatibility  [dataflake]


2.0.2 (2017-02-14)
------------------

- Fixed reflective XSS in findResult.
  This applies PloneHotfix20170117.  [maurits]


2.0.1 (2016-09-08)
------------------

- Quote variable in manage_tabs to avoid XSS.
  From Products.PloneHotfix20160830.  [maurits]


2.0.0 (2015-09-09)
------------------

- Moved code to https://github.com/zopefoundation/Products.ExternalEditor

- Update dtml to Zope trunk.


1.1.0 (2010-12-01)
------------------

- Added support for unaware mimetype browser - we now add the .zem extension no
  matter what the user-agent

- Modified the cache's parameters - special case for MSIE

- Corrected and added tests

- Moved the sources of the client to another package : collective.zopeedit


1.0 (2010-07-01)
----------------

- Update manage_main, manage_tabs, and findResult monkey patches to include
  fixes from Zope 2.12.


1.0a2 (2009-11-13)
------------------

- Removed imports from Globals.

- Purged old Zope 2 Interface interfaces for Zope 2.12 compatibility.


1.0a1 (2008-03-05)
------------------

- Updated package metadata to be usable as a package.


01/03/2007 - 0.9.3
------------------

- Fixed issue with 'manage_FTPget' overriding the 'Content-Type'
  header.

- Only run ExpandEnvironmentStrings on win32 systems.


9/14/2006 - 0.9.2
-----------------

- Added 'skip_data' option to make External Editor send out only
  the metadata part and skip appending data to the 'body'.

- Add a simple callback registry that can be used to add extra
  metadata headers or set special response headers when a file is
  edited through External Editor.

- Use rfc822.Message for parsing the metadata of the file being
  edited.

- Don't emit a warning about deprecated 'methods' in Zope >= 2.10.

- Fixed acquisition issue in manage_main.dtml to sync up with the
  same fix applied to Zope.


6/23/2005 - 0.9.1
-----------------

- Older pyc files for plugins were included in the 0.9 release.
  This version has the most up to date plugins.


6/20/2005 - 0.9
---------------

- When using the Excel plugin, errors were seen by users like
  "TypeError: This object does not support enumeration".  We now
  make the user deal with these.

- When using the Excel plugin, errors were intermittently raised to the
  user in the form "Fatal error: <unknown>.Path" and the user could
  subsequently not save the document back to Zope because the
  external editor process had quit.

- Changes to documents intermittently may not have been saved back to Zope
  when using any plugin that involved COM (Word, Excel, Powerpoint, etc).

- If Word was exited before a user actively saved, if there were
  outstanding changes in the document being edited, those changes
  would not be saved to the server despite the user answering
  "yes" to the "do you want to save before you exit" dialog
  presented by Word.

- The "title" attribute of the object being externally edited is
  now available within the set of headers returned by EE to the
  zopeedit client.

- Detecting whether the client has External Editor installed from
  within IE using JavaScript or VBScript is now possible, assuming
  that the client software is installed via the InnoSetup
  installer.  See "win32/ocx.txt" for more info.

- External Editor now compatible with objects that return a
  "filestream iterator" in Zope 2.7.1+. (if upgrading: this fix
  does not require an update to EE client, just the EE Zope
  product).

- Properly escape hyphens in man page. Thanks to Federico Sevilla III.

- Check if the editor was launched before locking the file in Zope. This
  prevents errors if the editor is closed before the lock request
  completes.

- Do not ask the user what editor to use on Posix platforms. Instead just
  tell the user to edit the config file. The askstring()
  function does not work with a hidden root Tk window in Python 2.3.4.
  Thanks to Christopher Mann.


7/13/04 - 0.8
-------------

- Add external editor icon to ZMI breadcrumbs for editable objects.

- Compiled windows helper app binary using Python 2.3.4, Pythonwin build
  163 and py2exe 0.5.

- Add Dreamweaver plugin contributed by Manuel Aristarann. Thanks also
  to Anton Stonor.

- Add ZMI support for Zope 2.7's ordered folder objects.

- Fix bug detecting basic auth info from older versions of CookieCrumber.
  Thanks to David D. Smith and Federico Sevilla III.

- Workaround IE browser error when running over SSL. Thanks to
  Marc-Aurele Darche.

- Add ".zem" file extension support for MacOS X (especially Mac IE),
  to ease helper app integration where MIME support is lacking. Thanks
  to Zac Bir.


zopeedit.py
#!/usr/local/bin/python
##############################################################################
#
# Copyright (c) 2001, 2002 Zope Corporation and Contributors.
# All Rights Reserved.
# 
# This software is subject to the provisions of the Zope Public License,
# Version 2.0 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
# 
##############################################################################

# Zope External Editor Helper Application by Casey Duncan

__version__ = '0.1'

import sys, os, stat
from time import sleep
from ConfigParser import ConfigParser
from httplib import HTTPConnection, HTTPSConnection

class Configuration:
    
    def __init__(self, path):
        # Create/read config file on instantiation
        self.path = path
        if not os.path.exists(path):
            f = open(path, 'w')
            f.write(default_configuration)
            f.close()
        self.changed = 0
        self.config = ConfigParser()
        self.config.readfp(open(path))
        
        
    def __del__(self):
        # Save changes on destruction
        if self.changed:
            self.save()
            
    def save(self):
        """Save config options to disk"""
        self.config.write(open(self.path, 'w'))
        self.changed = 0
        
            
    def set(self, section, option, value):
        self.config.set(section, option, value)
        self.changed = 1
    
    def __getattr__(self, name):
        # Delegate to the ConfigParser instance
        return getattr(self.config, name)
        
    def getAllOptions(self, meta_type, content_type):
        """Return a dict of all applicable options for the
           given meta_type and content_type
        """
        opt = {}
        sep = content_type.find('/')
        general_type = '%s/*' % content_type[:sep]
        sections = ('general', 
                    'meta-type:%s' % meta_type,
                    'content-type:%s' % general_type,
                    'content-type:%s' % content_type)
        for section in sections:
            if self.config.has_section(section):
                for option in self.config.options(section):
                    opt[option] = self.config.get(section, option)
        return opt
                     
        
class ExternalEditor:
    
    saved = 1
    
    def __init__(self, input_file):
        try:
            # Read the configuration file
            config_path = os.path.expanduser('~/.zope-external-edit')
            self.config = Configuration(config_path)

            # Open the input file and read the metadata headers
            in_f = open(input_file, 'rb')
            metadata = {}

            while 1:
                line = in_f.readline()[:-1]
                if not line: break
                sep = line.find(':')
                key = line[:sep]
                val = line[sep+1:]
                metadata[key] = val
            self.metadata = metadata

            self.options = self.config.getAllOptions(metadata['meta_type'],
                               metadata.get('content_type',''))

            # Write the body of the input file to a separate file
            body_file = self.metadata['url'][7:].replace('/',',')
            body_file = '%s-%s' % (os.tmpnam(), body_file)
            ext = self.options.get('extension')
            if ext and not body_file.endswith(ext):
                body_file = body_file + ext
            body_f = open(body_file, 'wb')
            body_f.write(in_f.read())
            in_f.close()
            body_f.close()
            self.clean_up = int(self.options.get('cleanup_files', 1))
            if self.clean_up: os.remove(input_file)
            self.body_file = body_file
        except:
            # for security, always delete the input file even if
            # a fatal error occurs, unless explicitly stated otherwise
            # in the config file
            if getattr(self, 'clean_up', 1):
                os.remove(input_file)
            raise
        
        # Get the host and path from the URL
        self.ssl = self.metadata['url'].startswith('https:')
        url = self.metadata['url'][7:]
        path_start = url.find('/')
        if path_start == -1:
            self.host = url
            self.path = '/'
        else:
            self.host = url[:path_start]
            self.path = url[path_start:]
        
    def __del__(self):
        # for security we always delete the files by default
        if getattr(self, 'clean_up', 1):
            os.remove(self.body_file)
            
    def getEditorPath(self):
        """Return the absolute path to the editor"""
        path = self.options.get('editor')
        
        if path:
            path = whereIs(path)
        elif has_tk():
            from tkSimpleDialog import askstring
            path = askstring('Zope External Editor', 
                             'Enter the default editor name')
            if not path: sys.exit(0)
            path = whereIs(path)
            if os.path.exists(path):
                self.config.set('general', 'editor', path)
                self.config.save()
                
        if not path:
            # Try some sensible default
            if sys.platform == 'win32':
                path = whereIs('notepad')
            else:
                editors = ['xedit', 'gvim', 'emacs']
                while not path:
                    path = whereIs(editors.pop())
                
        if path is not None:            
            return path
        else:
            self.fatalError('Editor not found at "%s"' % path)
        
    def launch(self):
        """Launch external editor"""
        editor = self.getEditorPath()
        file = self.body_file
        last_fstat = os.stat(file)
        pid = os.spawnl(os.P_NOWAIT, editor, editor, file)
        use_locks = int(self.options.get('use_locks'))
        
        if use_locks:
            self.lock()
            
        exit_pid = 0
        save_interval = self.config.getfloat('general', 'save_interval')
        launched = 0
        
        while exit_pid != pid:
            sleep(save_interval or 2)
            
            try:
                exit_pid, exit_status = os.waitpid(pid, os.WNOHANG)
                if not exit_pid: launched = 1
            except OSError:
                exit_pid = pid
            
            fstat = os.stat(file)
            if (exit_pid == pid or save_interval) \
               and fstat[stat.ST_MTIME] != last_fstat[stat.ST_MTIME]:
                # File was modified
                self.saved = self.putChanges()
                last_fstat = fstat
                
        if not launched:
            fatalError(('Editor "%s" did not launch.\n'
                        'It may not be a graphical editor.') % editor)
         
        if use_locks:
            self.unlock()
        
        if not self.saved and has_tk():
            from tkMessageBox import askyesno
            if askyesno('Zope External Editor',
                        'File not saved to Zope.\nReopen local copy?'):
                has_tk() # ugh, keeps tk happy
                self.launch()
            else:
                self.clean_up = 0 # Keep temp file
                has_tk() # ugh
        
    def putChanges(self):
        """Save changes to the file back to Zope"""
        f = open(self.body_file, 'rb')
        body = f.read()
        f.close()
        headers = {'Content-Type': 
                   self.metadata.get('content_type', 'text/plain')}
        
        if hasattr(self, 'lock_token'):
            headers['If'] = '<%s> (<%s>)' % (self.path,
                                             self.lock_token)
        
        response = self.zope_request('PUT', headers, body)
        del body # Don't keep the body around longer then we need to

        if response.status / 100 != 2:
            # Something went wrong
            sys.stderr.write('Error occurred during HTTP put:\n%d %s\n' \
                             % (response.status, response.reason))
            sys.stderr.write('\n----\n%s\n----\n' % response.read())
            
            message = response.getheader('Bobo-Exception-Type')
            if has_tk():
                from tkMessageBox import askretrycancel
                if askretrycancel('Zope External Editor',
                                  ('Could not save to Zope.\nError '
                                   'occurred during HTTP put:\n%d %s\n%s') \
                                  % (response.status, response.reason,
                                     message)):
                    has_tk() # ugh, keeps tk happy
                    self.putChanges()
                else:
                    has_tk() # ugh
                    return 0
        return 1
    
    def lock(self):
        """Apply a webdav lock to the object in Zope"""
        
        headers = {'Content-Type':'text/xml; charset="utf-8"',
                   'Timeout':'infinite',
                   'Depth':'infinity',
                  }
        body = ('<?xml version="1.0" encoding="utf-8"?>\n'
                '<d:lockinfo xmlns:d="DAV:">\n'
                '  <d:lockscope><d:exclusive/></d:lockscope>\n'
                '  <d:locktype><d:write/></d:locktype>\n'
                '  <d:depth>infinity</d:depth>\n'
                '  <d:owner>\n' 
                '  <d:href>Zope External Editor</d:href>\n'
                '  </d:owner>\n'
                '</d:lockinfo>'
                )
        
        response = self.zope_request('LOCK', headers, body)
        
        if response.status / 100 == 2:
            # We got our lock, extract the lock token and return it
            reply = response.read()
            token_start = reply.find('>opaquelocktoken:')
            token_end = reply.find('<', token_start)
            if token_start > 0 and token_end > 0:
               self.lock_token = reply[token_start+1:token_end]
        else:
            # We can't lock her sir!
            if response.status == 423:
                message = '\n(Object already locked)'
            else:
                message = response.getheader('Bobo-Exception-Type')

            sys.stderr.write('Error occurred during lock request:\n%d %s\n' \
                             % (response.status, response.reason))
            sys.stderr.write('\n----\n%s\n----\n' % response.read())
            if has_tk():
                from tkMessageBox import askretrycancel
                if askretrycancel('Zope External Editor',
                                  ('Lock request failed:\n%d %s\n%s') \
                                  % (response.status, response.reason, message)):
                    has_tk() # ugh, keeps tk happy
                    return self.lock()
                else:
                    has_tk() # ugh
                    return 0
        return 1
                    
    def unlock(self):
        """Remove webdav lock from edited zope object"""
        if not hasattr(self, 'lock_token'): 
            return 0
            
        headers = {'Lock-Token':self.lock_token}
        
        response = self.zope_request('UNLOCK', headers)
        
        if response.status / 100 != 2:
            # Captain, she's still locked!
            message = response.getheader('Bobo-Exception-Type')
            sys.stderr.write('Error occurred during unlock request:\n%d %s\n%s\n' \
                             % (response.status, response.reason, message))
            sys.stderr.write('\n----\n%s\n----\n' % response.read())
            if has_tk():
                from tkMessageBox import askretrycancel
                if askretrycancel('Zope External Editor',
                                  ('Unlock request failed:\n%d %s') \
                                  % (response.status, response.reason)):
                    has_tk() # ugh, keeps tk happy
                    return self.unlock(token)
                else:
                    has_tk() # ugh
                    return 0
        return 1
        
    def zope_request(self, method, headers={}, body=''):
        """Send a request back to Zope"""
        try:
            if self.ssl:
                h = HTTPSConnection(self.host)
            else:
                h = HTTPConnection(self.host)

            h.putrequest(method, self.path)
            #h.putheader("Host", self.host)  # required by HTTP/1.1
            h.putheader('User-Agent', 'Zope External Editor/%s' % __version__)
            h.putheader('Connection', 'close')

            for header, value in headers.items():
                h.putheader(header, value)

            h.putheader("Content-Length", str(len(body)))

            if self.metadata.get('auth','').startswith('Basic'):
                h.putheader("Authorization", self.metadata['auth'])

            if self.metadata.get('cookie'):
                h.putheader("Cookie", self.metadata['cookie'])

            h.endheaders()
            h.send(body)
            return h.getresponse()
        except:
            # On error return a null response with error info
            class NullResponse:
                def getheader(n,d): return d
                def read(self): return '(No Response From Server)'
            
            response = NullResponse()
            response.reason = sys.exc_info()[1]
            try:
                response.status, response.reason = response.reason
            except:
                response.status = 0
            return response

def has_tk():
    """Sets up a suitable tk root window if one has not
       already been setup. Returns true if tk is happy,
       false if tk throws an error (like its not available)"""
    if not hasattr(globals(), 'tk_root'):
        # create a hidden root window to make Tk happy
        try:
            global tk_root
            from Tkinter import Tk
            tk_root = Tk()
            tk_root.withdraw()
            return 1
        except:
            return 0
    return 1
        
def fatalError(message, exit=1):
    """Show error message and exit"""
    message = 'FATAL ERROR:\n%s\n' % message
    try:
        if has_tk():
            from tkMessageBox import showerror
            showerror('Zope External Editor', message)
            has_tk()
    finally:
        sys.stderr.write(message)
        if exit: sys.exit(0)

def whereIs(filename): 
    """Given a filename, returns the full path to it based
       on the PATH environment variable. If no file was found, None is returned
    """
    pathExists = os.path.exists
    
    if pathExists(filename):
        return filename
        
    pathJoin = os.path.join
    paths = os.environ['PATH'].split(os.pathsep)
    
    for path in paths:
        file_path = pathJoin(path, filename)
        if pathExists(file_path):
            return file_path

default_configuration = """\
# Zope External Editor helper application configuration

[general]
# General configuration options

# Uncomment and specify an editor value to override the editor
# specified in the environment
#editor = 

# Automatic save interval, in seconds. Set to zero for
# no auto save (save to Zope only on exit).
save_interval = 1

# Temporary file cleanup. Set to false for debugging or
# to waste disk space. Note: setting this to false is a
# security risk to the zope server
cleanup_files = 1

# Use WebDAV locking to prevent concurrent editing by
# different users. Disable for single user use for
# better performance
use_locks = 1

# Specific settings by content-type or meta-type. Specific
# settings override general options above. Content-type settings
# override meta-type settings for the same option.

[meta-type:DTML Document]
extension=.dtml

[meta-type:DTML Method]
extension=.dtml

[meta-type:Script (Python)]
extension=.py

[meta-type:Page Template]
extension=.pt

[meta-type:Z SQL Method]
extension=.sql

[content-type:text/*]
extension=.txt

[content-type:text/html]
extension=.html

[content-type:text/xml]
extension=.xml

[content-type:image/*]
editor=gimp

[content-type:image/gif]
extension=.gif

[content-type:image/jpeg]
extension=.jpg

[content-type:image/png]
extension=.png"""

if __name__ == '__main__':
    try:
        ExternalEditor(sys.argv[1]).launch()
    except KeyboardInterrupt:
        pass
    except SystemExit:
        pass
    except:
        fatalError(sys.exc_info()[1],0)
        raise
